src/client.ts

```diff
@@ -1179,57 +1197,77 @@ export class ZkBobClient {
     // no parts when no requests
     if (transfers.length == 0) return [];
 
-    let result: Array<TransferConfig> = [];
-    let txNotes: Array<TransferRequest> = [];
-    const accountBalance = await state.accountBalance();
-    const notesParts = await this.getGroupedNotes(tokenAddress);
+    let aggregatedTransfers: MultinoteTransferRequest[] = [];
+    for (let i = 0; i < transfers.length; i += CONSTANTS.OUT) {
+      const requests = transfers.slice(i, i + CONSTANTS.OUT);
+      const totalAmount = requests.reduce(
+        (acc, cur) => acc + cur.amountGwei,
+        BigInt(0)
+      );
+      aggregatedTransfers.push({totalAmount, requests});
+    }
+
+    let accountBalance: bigint = await state.accountBalance();
+    const groupedNotesBalances = await this.getGroupedNotes(tokenAddress);
 
-    let requestIdx = 0;
-    let txIdx = 0;
-    let remainRequestAmount = transfers[requestIdx].amountGwei;
-    let oneTxPart = accountBalance;
+    let aggregationParts: Array<TransferConfig> = [];
+    let txParts: Array<TransferConfig> = [];
+    
+    let i = 0;
     do {
-      if (txIdx < notesParts.length) {
-        oneTxPart += notesParts[txIdx];
+      txParts = this.tryToPrepareTransfers(accountBalance, feeGwei, groupedNotesBalances.slice(i, i + aggregatedTransfers.length), aggregatedTransfers);
+      if (txParts.length == aggregatedTransfers.length) {
+        // We are able to perform all txs starting from this index
+        return aggregationParts.concat(txParts);
       }
 
-      oneTxPart -= feeGwei; // available token amount for tx (account + notes)
-      // create output notes for the transaction
-      while(oneTxPart > 0 && requestIdx < transfers.length && txNotes.length < CONSTANTS.OUT) {
-        let noteAmount = oneTxPart;
-        if (oneTxPart > remainRequestAmount) {
-          noteAmount = remainRequestAmount;
-        }
-        
-        // add output note for the current transaction
-        txNotes.push({amountGwei: noteAmount, destination: transfers[requestIdx].destination});
-
-        oneTxPart -= noteAmount;
-        remainRequestAmount -= noteAmount;
-        if(remainRequestAmount == BigInt(0)) {
-          if (++requestIdx < transfers.length) {
-            remainRequestAmount = transfers[requestIdx].amountGwei;
-          }
-        }
+      if (groupedNotesBalances.length == 0) {
```

Comment:
should it be this way, or later `groupedNotesBalances[i]` will be undefined?
```suggestion
      if (groupedNotesBalances.length <= i) {
```

Reply:
`i` can be equal to `groupedNotesBalances.length` only on the first iteration and only if `groupedNotesBalances.length == 0` since we are checking that `i < groupedNotesBalances.length` in the while. So it seems to me that these two conditions are equivalent.